src/R2N.jl

```diff
@@ -0,0 +1,387 @@
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+  non_mono_size = 1,
+  kwargs...,
+) where {T, V}
+  solver = R2NSolver(nlp; non_mono_size = non_mono_size, subsolver_type = subsolver_type)
+  return solve!(solver, nlp; non_mono_size = non_mono_size, kwargs...)
+end
+
+function SolverCore.solve!(
+  solver::R2NSolver{T, V},
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = nlp.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  η1 = T(0.0001),
+  η2 = T(0.001),
+  λ = T(2), #>1
+  σmin = zero(T),# μmin = σmin to match the paper
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  max_iter::Int = typemax(Int),
+  verbose::Int = 0,
+  subsolver_verbose::Int = 0,
+  non_mono_size = 1,
+) where {T, V}
+  unconstrained(nlp) || error("R2N should only be called on unconstrained problems.")
+  if non_mono_size < 1
+    error("non_mono_size must be greater than or equal to 1")
+  end
+  if (solver.subsolver_type isa ShiftedLBFGSSolver && !isa(nlp, LBFGSModel))
+    error("Unsupported subsolver type, ShiftedLBFGSSolver is only can be used by LBFGSModel")
+  end
+  @assert(λ > 1) #λ must be greater than 1
+  reset!(stats)
+  start_time = time()
+  set_time!(stats, 0.0)
+  μmin = σmin
+  n = nlp.meta.nvar
+  x = solver.x .= x
+  ck = solver.cx
+  ∇fk = solver.gx # k-1
+  ∇fn = solver.gn #current 
+  s = solver.s
+  H = solver.H
+  shifted_H = solver.shifted_H
+  Hs = solver.Hs
+  σk = solver.σ
+  μk = solver.μ
+  cgtol = solver.cgtol
+
+  set_iter!(stats, 0)
+  f0 = obj(nlp, x)
+  set_objective!(stats, f0)
+
+  grad!(nlp, x, ∇fk)
+  isa(nlp, QuasiNewtonModel) && (∇fn .= ∇fk)
+  norm_∇fk = norm(∇fk)
+  set_dual_residual!(stats, norm_∇fk)
+
+  μk = 2^round(log2(norm_∇fk + 1)) / norm_∇fk
+  σk = μk * norm_∇fk
+  ρk = zero(T)
+
+  # Stopping criterion: 
+  fmin = min(-one(T), f0) / eps(T)
+  unbounded = f0 < fmin
+  
+  ϵ = atol + rtol * norm_∇fk
+  optimal = norm_∇fk ≤ ϵ
+
+  if optimal
+    @info("Optimal point found at initial point")
+    @info @sprintf "%5s  %9s  %7s  %7s  %7s  %7s  %1s" "iter" "f" "‖∇f‖" "μ" "σ" "ρ" ""
+    @info @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e  %+7.1e  %1s" stats.iter stats.objective norm_∇fk μk σk ρk ""
+  end
+  if verbose > 0 && mod(stats.iter, verbose) == 0
+    @info @sprintf "%5s  %9s  %7s  %7s  %7s  %7s  %1s" "iter" "f" "‖∇f‖" "μ" "σ" "ρ" ""
+    infoline =
+      @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e  %+7.1e  %1s" stats.iter stats.objective norm_∇fk μk σk ρk ""
+  end
+
+  set_status!(
+    stats,
+    get_status(
+      nlp,
+      elapsed_time = stats.elapsed_time,
+      optimal = optimal,
+      unbounded = unbounded,
+      max_eval = max_eval,
+      iter = stats.iter,
+      max_iter = max_iter,
+      max_time = max_time,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+  cgtol = max(rtol, min(T(0.1), √norm_∇fk, T(0.9) * cgtol))
+  
+  local ρk
```

Comment:
?

Reply:
As suggested by @dpo since we have allocation there